### PR TITLE
feat(#2544): add API for querying win ID, api.tree.winid()

### LINKED ISSUE #2544: api.tree.winid()
### Discussed in https://github.com/nvim-tree/nvim-tree.lua/discussions/2532

<div type='discussions-op-text'>

<sup>Originally posted by **j-hui** November 14, 2023</sup>
I'm trying to see if I can get Fidget to avoid rendering on top of the nvim-tree window when it's open. I see there's already an event subscription mechanism, but is there any way to dynamically query the position and width of the nvim-tree window?

I saw [this question](https://github.com/nvim-tree/nvim-tree.lua/discussions/1813) but it suggests the only option is to enumerate through all window IDs and search for the nvim-tree window. Is there any API that is more direct?

Thanks!</div>

### PR DESCRIPTION
**Pull request by j-hui**
Closes #2544

### CODE REVIEW

File: doc/nvim-tree-lua.txt [hunk 1]
```diff
@@ -1698,6 +1698,18 @@ tree.is_visible({opts})                      *nvim-tree-api.tree.is_visible()*
...
       (boolean) nvim-tree is visible
 
+tree.winid({opts})                              *nvim-tree-api.tree.winid()*
+    Retrieve the winid of the open tree.
+
+    Parameters: ~
+      • {opts} (table) optional parameters
+
+    Options: ~
+      • {tabpage}        (number)  tabpage, 0 for current, default 0
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:1708]:
```suggestion
      • {tabpage}        (number|nil)  tabpage, 0 or nil for current, default nil
```

Apologies, I did not follow the semantics of `is_tree_buf`

File: lua/nvim-tree/api.lua [hunk 2]
```diff
@@ -145,6 +145,8 @@ Api.tree.is_tree_buf = wrap(require("nvim-tree.utils").is_nvim_tree_buf)
 
 Api.tree.is_visible = wrap(require("nvim-tree.view").is_visible)
 
+Api.tree.winid = wrap(require("nvim-tree.view").winid)
```
Comment (alex-courtis) [inline lua/nvim-tree/api.lua:148]:
```suggestion
---@class ApiTreeWinIdOpts
---@field tabpage number|nil default nil

Api.tree.winid = wrap(require("nvim-tree.view").winid)
```

File: lua/nvim-tree/view.lua [hunk 3]
```diff
@@ -427,6 +427,17 @@ function M.focus(winnr, open_if_closed)
   vim.api.nvim_set_current_win(wnr)
 end
 
+--- Retrieve the winid of the open tree.
+--- Unlike get_winnr(), this returns nil if the nvim-tree window is not visible.
```
Comment (alex-courtis) [inline lua/nvim-tree/view.lua:431]:
```suggestion
--- @param opts ApiTreeWinIdOpts|nil
--- @return number|nil winid unlike get_winnr(), this returns nil if the nvim-tree window is not visible
```

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looking great! One failure with 0 tabpage.

Testing:

no tabpage:
- [x] startup - nil
- [x] tree open - valid
- [x] tree closed - nil
- [x] create tab 2 - nil
- [x] tab 2 tree open - valid
- [x] tab 1 tree open - valid
- [x] tab 2 tree closed - nil

tabpage 1 provided:
- [x] startup - nil
- [x] tree open - valid
- [x] tree closed - nil
- [x] create tab 2 - nil
- [x] tab 2 tree open - nil
- [x] tab 1 tree open from tab 2 - valid
- [x] tab 2 tree closed - valid
- [x] tab 2 and tab 1 tree closed - nil

tabpage 0 provided:
- [ ] always nil

**Comment by j-hui** [comment]
Thanks for your quick and detailed feedback! I just pushed a change to address that issue you found when `tabpage` is 0.

**Review by alex-courtis — APPROVED** [review]
Works nicely for 0 and nil; many thanks for your contribution!

**Pull request merged**